File: coordinator/src/coord/testCoordinator.go
package coordinator 

import "log"
import "fmt"

type TestCoord struct {
	currentTask TaskID
	seed int64
	tasksFinished int
  results []int
  tasks map[TaskID]int
  numSubTasks int
  result int
}

func (sc *TestCoord) Init(co *Coordinator, seed int64) {
  sc.tasks = make(map[TaskID]int)
  sc.results = make([]int, 0)
  sc.numSubTasks = 400
	// Start a task
  for i := 0; i < sc.numSubTasks; i++ {
    s := i * 4
    args := []int{s, s + 1, s + 2, s + 3}
    sc.goPy(co, "sum", args, 1)
  }
}

func (sc *TestCoord) ClientJoined(co *Coordinator, CID ClientID, NumNodes int) {
	// Do nothing for now 
	fmt.Printf("Client joined: %v\n", CID)
}

func (sc *TestCoord) ClientDead(co *Coordinator, CID ClientID) {
	// Do nothing for now 
	fmt.Printf("Client marked dead: %v\n", CID)
}

func (sc *TestCoord) goPy(co *Coordinator, name string, args []int, taskType int) {
  params := TaskParams{}
  params.FuncName = name
  params.DoneKeys = []string{"result"}
  params.BaseObject = args
  tid := co.StartTask(params)
  sc.tasks[tid] = taskType
}

func (sc *TestCoord) TaskDone(co *Coordinator,
  TID TaskID, DoneValues map[string]interface{}) {
	// Handle tasks with prerequisites here 
  v, p := sc.tasks[TID]
  if !p {
    log.Fatal("TID was never created.")
  }

  result := int(DoneValues["result"].(float64))
  if v == 1 {
    sc.results = append(sc.results, result)

    if len(sc.results) == sc.numSubTasks {
      sc.goPy(co, "sum", sc.results, 2)
    }
	} else if v == 2 {
    /* fmt.Printf("Received: %d\n", result) */
    co.Finish([]TaskID{TID})
    sc.result = result
  }
}

// Returns 0 if not finished, > 0 when done
func (sc *TestCoord) Result() int {
  return sc.result
}

func MakeTestCoord() *TestCoord {
	sc := new(TestCoord)
	return sc
}

File: libraries/python/coordinator/testCoordinator.py
#!/usr/bin/python
from TaskSprintCoordinator import *

class TestCoordinator(TaskSprintCoordinator):
  def init(self, seed):
    alltasks = []
    num = 200
    for i in xrange(num):
      tid = self.start_task(
        name = "sum",
        base = [i * (num / 10) + k for k in range(num / 10)],
      )
      alltasks.append(tid)

    finaltasks = []
    toassign, size = num, num / 10
    for i in xrange(toassign / size):
      pretasks = alltasks[i * size : (i + 1) * size]
      print pretasks
      tid = self.start_task(
        name = "sum",
        prekeys = ["result" for i in range(size)],
        pretasks = pretasks
      )
      finaltasks.append(tid)

    self.final = self.start_task(
      name = "sum",
      prekeys = ["result" for i in range(len(finaltasks))],
      pretasks = finaltasks,
      keys = ["result"]
    )

  def client_joined(self, cid, num_nodes):
    print "Somebody joined! %d" %cid

  def client_dead(self, cid):
    print "Somebody died :( %d" %cid

  def task_done(self, tid, values):
    if tid == self.final:
      self.finish(taskids = [tid], values = values)
    print "Task %d is done. Result: %s" %(tid, values)

TestCoordinator().start()
